File: 0133-clone-graph/0133-clone-graph.py
from collections import deque
from typing import Optional


class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node: return None

        nodeDict = dict()
        clone = Node(val=node.val)
        nodeDict[clone.val] = clone
        que = deque([node])

        while que:
            chk = que.popleft()

            for nb in chk.neighbors:
                if nb.val not in nodeDict:
                    new = Node(val=nb.val)
                    nodeDict[nb.val] = new
                    nodeDict[chk.val].neighbors.append(new)
                    que.append(nb)
                else:
                    target = nodeDict[nb.val]
                    if target not in nodeDict[chk.val].neighbors:
                        nodeDict[chk.val].neighbors.append(target)

        return nodeDict[node.val]

File: 0133-clone-graph/0133-clone-graph.java
class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) {
            return null;
        }

        Map<Integer, Node> nodeMap = new HashMap<>();
        Queue<Node> que = new ArrayDeque<>();
        Node clone = new Node(node.val);
        nodeMap.put(clone.val, clone);
        que.add(node);

        while (!que.isEmpty()) {
            Node now = que.poll();

            for (Node nb: now.neighbors) {
                if (!nodeMap.containsKey(nb.val)) {
                    Node newNode = new Node(nb.val);
                    nodeMap.put(newNode.val, newNode);
                    List<Node> temp = nodeMap.get(now.val).neighbors;
                    temp.add(newNode);
                    que.add(nb);
                } else {
                    Node target = nodeMap.get(nb.val);
                    List<Node> temp = nodeMap.get(now.val).neighbors;
                    temp.add(target);
                }
            }
        }

        return nodeMap.get(node.val);
    }
}